adapters/tappx/tappx.go

```diff
@@ -0,0 +1,207 @@
+package tappx
+
+import (
+	"encoding/json"
+	"fmt"
+	"text/template"
+	//"strings"
+	"github.com/golang/glog"
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/macros"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	//"github.com/prebid/prebid-server/config"
+	"net/http"
+	"strconv"
+	"time"
+)
+
+const TAPPX_BIDDER_VERSION = "1.0"
+
+type TappxAdapter struct {
+	http             *adapters.HTTPAdapter
+	endpointTemplate template.Template
+}
+
+func NewTappxAdapter(config *adapters.HTTPAdapterConfig, endpointTemplate string) adapters.Bidder {
+	return NewTappxBidder(adapters.NewHTTPAdapter(config).Client, endpointTemplate)
+}
+
+func NewTappxBidder(client *http.Client, endpointTemplate string) *TappxAdapter {
+	a := &adapters.HTTPAdapter{Client: client}
+	template, err := template.New("endpointTemplate").Parse(endpointTemplate)
+	if err != nil {
+		glog.Fatal("Unable to parse endpoint url template")
+		return nil
+	}
+	return &TappxAdapter{
+		http:             a,
+		endpointTemplate: *template,
+	}
+}
+
+type tappxParams struct {
+	Host     string `json:"host"`
+	TappxKey string `json:"tappxkey"`
+	Endpoint string `json:"endpoint"`
+}
+
+func (a *TappxAdapter) Name() string {
+	return "tappx"
+}
+
+func (a *TappxAdapter) SkipNoCookies() bool {
+	return false
+}
+
+func (a *TappxAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	if len(request.Imp) == 0 {
+		return nil, []error{&errortypes.BadInput{
+			Message: "No impression in the bid request",
+		}}
+	}
+
+	errs := make([]error, 0, len(request.Imp))
+
+	var bidderExt adapters.ExtImpBidder
```

Comment:
Are we sure the `request.Imp[]` array will always come with one element? If not, do we only care for the first one? Is there any possibility of a second or third `Imp[]` element to be relevant?

Reply:
Most of our traffic comes with only one impression and we only validate the first Imp object. If that first element contains some error we will discard the request. If many Imp objects are sent and there is some error in the other objects we handle this directly in our SSP.